AppTaskQueue/appscale/taskqueue/service_manager.py

```diff
@@ -0,0 +1,154 @@
+    """ Creates a new ProjectServiceManager.
+
+    Args:
+      zk_client: A KazooClient.
+      db_access: A DatastoreProxy.
+      project_id: A string specifying a project ID.
+    """
+    super(ProjectServiceManager, self).__init__()
+    self.zk_client = zk_client
+    self.db_access = db_access
+    self.project_id = project_id
+    services_node = '/appscale/projects/{}/services'.format(self.project_id)
+    zk_client.ensure_path(services_node)
+    self.watch = zk_client.ChildrenWatch(services_node,
+                                         self._update_services_watch)
+
+  def update_services(self, new_services_list):
+    """ Establishes watches for all existing services.
+
+    Args:
+      new_services_list: A list of strings specifying existing service IDs.
+    """
+    to_remove = [service for service in self if service not in
+                 new_services_list]
+    for service_id in to_remove:
+      logger.debug("updating: {}".format(to_remove))
+      self[service_id].stop()
+      del self[service_id]
+
+    for service_id in new_services_list:
+      logger.debug("updating: {}".format(new_services_list))
+      if service_id not in self:
+        self[service_id] = VersionPortManager(self.zk_client, self.db_access,
+                                              self.project_id, service_id)
+
+  def _update_services_watch(self, new_services):
+    """ Handles creation and deletion of services.
+
+    Since this runs in a separate thread, it doesn't change any state directly.
+    Instead, it just acts as a bridge back to the main IO loop.
+
+    Args:
+      new_services: A list of strings specifying all existing services.
+    """
+    main_io_loop = IOLoop.instance()
+    main_io_loop.add_callback(self.update_services, new_services)
+
+class VersionPortManager(dict):
+  """ Keeps track of version port details for a single service. """
+  def __init__(self, zk_client, db_access, project_id, service_id):
+    """ Creates a new VersionPortManager.
+
+    Args:
+      zk_client: A KazooClient.
+      db_access: A DatastoreProxy.
+      project_id: A string specifying a project ID.
+    """
+    super(VersionPortManager, self).__init__()
+    self.zk_client = zk_client
+    self.db_access = db_access
+    self.versions_node = '/appscale/projects/{0}/services/{1}/versions'.format(
+      project_id, service_id)
+    zk_client.ensure_path(self.versions_node)
+    self.watch = zk_client.ChildrenWatch(self.versions_node,
+                                         self._update_versions_watch)
+
+  def update_version_ports(self, new_versions_list):
+    """ Establishes watches for all existing versions and ports.
+
+    Args:
+      new_versions_list: A list of strings specifying existing version IDs.
+    """
+    to_remove = [service for service in self if service not in
+                 new_versions_list]
+    for version_id in to_remove:
+      logger.debug("updating: {}".format(to_remove))
+      self.watch.__stopped = True
```

Comment:
ok, so I should just remove this line? I was confused about this specific watch.